kratos/mpi/utilities/mpi_assemble_utilities.h

```diff
@@ -0,0 +1,305 @@
+            std::vector<int> all_node_ids;
+            if (my_rank == 0) {
+                for (const auto& rank : all_rank_node_ids) {
+                    for (const auto node_id : rank) {
+                        all_node_ids.push_back(node_id);
+                    }
+                }
+                std::sort(all_node_ids.begin(), all_node_ids.end());
+                auto last = std::unique(all_node_ids.begin(), all_node_ids.end());
+                all_node_ids.erase(last, all_node_ids.end());
+            }
+
+            int number_of_communication_nodes = all_node_ids.size();
+            // now we get all nodes which requires updating in all_node_ids in all ranks
+            r_data_communicator.Broadcast(number_of_communication_nodes, 0);
+            if (my_rank != 0) {
+                all_node_ids.resize(number_of_communication_nodes);
+            }
+            r_data_communicator.Broadcast(all_node_ids, 0);
+
+            // identify ranks which these nodes belongs to
+            auto& local_mesh = r_communicator.LocalMesh();
+            std::vector<int> local_node_id_ranks(number_of_communication_nodes);
+
+            IndexPartition<int>(number_of_communication_nodes).for_each([&](const int Index) {
+                local_node_id_ranks[Index] =
+                    local_mesh.HasNode(all_node_ids[Index]) ? my_rank : -1;
+            });
+
+            const auto& all_rank_nodal_id_ranks =
+                r_data_communicator.Gatherv(local_node_id_ranks, 0);
+            std::vector<int> all_node_id_ranks(number_of_communication_nodes, -1);
+            if (my_rank == 0) {
+                IndexPartition<int>(number_of_communication_nodes).for_each([&](const int Index) {
+                    for (unsigned int i_rank = 0;
+                         i_rank < all_rank_nodal_id_ranks.size(); ++i_rank) {
+                        const int node_rank = all_rank_nodal_id_ranks[i_rank][Index];
+                        if (node_rank > -1) {
+                            if (all_node_id_ranks[Index] == -1) {
+                                all_node_id_ranks[Index] = node_rank;
+                            } else if (all_node_id_ranks[Index] != node_rank) {
+                                KRATOS_ERROR
+                                    << "Node id " << all_node_ids[Index]
+                                    << " does belong to more than one rank, "
+                                    << "which is not allowed.\n";
+                            }
+                            break;
+                        }
+                    }
+
+                    KRATOS_ERROR_IF(all_node_id_ranks[Index] == -1)
+                        << "Node id " << all_node_ids[Index]
+                        << " not found in any of the ranks.\n";
+                });
+            }
+
+            // now broadcast all the ranks properly to all ranks
+            r_data_communicator.Broadcast(all_node_id_ranks, 0);
```

Comment:
ouch this is very expensive, a all to all communication. I don't have a smarter way of doing this, however it is definitely very expensive.